File: agent/internal/api/protocol.go
// Package api is the agent's HTTP + WebSocket control surface: the REST
// endpoints the panel calls (power actions, status) and the WebSocket the
// browser connects to DIRECTLY for console and stats (architecture doc
// 3.4 / 4.5 / 7).
package api

import "github.com/coder/websocket"

// Envelope is the wire format for every WebSocket message in both
// directions: {"event": "...", "data": {...}}. This mirrors the protocol
// documented in architecture doc 4.5/4.3 exactly, so a browser client and
// this agent implementation share one message shape.
type Envelope struct {
	Event string `json:"event"`
	Seq   uint64 `json:"seq,omitempty"`
	Data  any    `json:"data,omitempty"`
	TS    int64  `json:"ts,omitempty"`
}

// Inbound event names (client -> agent).
const (
	EventAuth         = "auth"
	EventConsoleSend  = "console:send"
	EventPowerSet     = "power:set"
	EventLogsRequest  = "logs:request"
	EventStatsRequest = "stats:request"
	EventPing         = "ping"
)

// Outbound event names (agent -> client).
const (
	EventAuthOK           = "auth:ok"
	EventConsoleOutput    = "console:output"
	EventConsoleTruncated = "console:truncated"
	EventStatus           = "status"
	EventStats            = "stats"
	EventTokenExpiring    = "token:expiring"
	EventTokenExpired     = "token:expired"
	EventError            = "error"
	EventPong             = "pong"
)

// WS close codes, matching architecture doc 4.3 exactly so a client can
// branch on them without parsing the reason string.
const (
	StatusAuthFailed       websocket.StatusCode = 4000
	StatusTokenExpired     websocket.StatusCode = 4001
	StatusPermissionDenied websocket.StatusCode = 4003
	StatusServerNotFound   websocket.StatusCode = 4004
	StatusServerSuspended  websocket.StatusCode = 4009
)

type authData struct {
	Token string `json:"token"`
}

type consoleSendData struct {
	Command string `json:"command"`
}

type powerSetData struct {
	Action string `json:"action"`
}

type authOKData struct {
	Permissions []string `json:"permissions"`
	ExpiresAt   int64    `json:"expiresAt"`
}

type consoleOutputData struct {
	Line   string `json:"line"`
	Stream string `json:"stream"`
}

type consoleTruncatedData struct {
	Dropped uint64 `json:"dropped"`
}

type statusData struct {
	State    string `json:"state"`
	Previous string `json:"previous,omitempty"`
}

type tokenExpiringData struct {
	SecondsRemaining int64 `json:"secondsRemaining"`
}

type errorData struct {
	Code    string `json:"code"`
	Message string `json:"message"`
	Fatal   bool   `json:"fatal"`
}

File: apps/panel/src/shared/realtime/protocol.ts
// Mirrors agent/internal/api/protocol.go exactly — one wire format shared
// by the browser and the agent, no translation layer in between
// (architecture doc 4.5/5.2).

export interface Envelope<T = unknown> {
  event: string;
  seq?: number;
  data?: T;
  ts?: number;
}

export const EventAuth = 'auth';
export const EventConsoleSend = 'console:send';
export const EventPowerSet = 'power:set';
export const EventPing = 'ping';

export const EventAuthOK = 'auth:ok';
export const EventConsoleOutput = 'console:output';
export const EventConsoleTruncated = 'console:truncated';
export const EventStatus = 'status';
export const EventStats = 'stats';
export const EventTokenExpiring = 'token:expiring';
export const EventTokenExpired = 'token:expired';
export const EventError = 'error';
export const EventPong = 'pong';

export const StatusAuthFailed = 4000;
export const StatusTokenExpired = 4001;
export const StatusPermissionDenied = 4003;
export const StatusServerNotFound = 4004;
export const StatusServerSuspended = 4009;

export interface AuthData {
  token: string;
}
export interface ConsoleSendData {
  command: string;
}
export interface PowerSetData {
  action: string;
}
export interface AuthOKData {
  permissions: string[];
  expiresAt: number;
}
export interface ConsoleOutputData {
  line: string;
  stream: string;
}
export interface StatusData {
  state: string;
  previous?: string;
}
export interface ErrorData {
  code: string;
  message: string;
  fatal: boolean;
}
export interface StatsFrame {
  state: string;
  cpu_percent: number;
  cpu_limit_percent: number;
  memory_bytes: number;
  memory_limit_bytes: number;
  disk_bytes: number;
  disk_limit_bytes: number;
  network_rx_bytes: number;
  network_tx_bytes: number;
  uptime_ms: number;
}
